fix: CheckBox

seitbnao opened the pull request:
Corrige o click nos check box adicionados nas fieldScenes, o erro basicamente é a frequência em que ocorre o click do mouse.

seitbnao committed:
fix: CheckBox

Corrige o click nos check box adicionados nas fieldScenes, o erro basicamente é a frequência em que ocorre o click do mouse.

kevinkouketsu commented:
Gostaria de entender um pouco melhor. 
Como posso ver o problema acontecendo? O que devo notar?

kevinkouketsu commented on Projects/TMProject/SControl.cpp:
```diff
@@ -1639,11 +1639,11 @@ int SCheckBox::OnMouseEvent(unsigned int dwFlags, unsigned int wParam, int nX, i
+	if (dwFlags == 512 || dwFlags == 513)
 		return 0;
 
-	if (dwFlags != 513)
```
Essa remoção não é necessária
512 refere-se a `WM_MOUSEMOVE`. Isto é, ignoramos movimentos de mouse.
513 refere-se a `WM_LBUTTONDOWN`, que é um clique do botão esquerdo.

Lendo o código, faz sentido que a gente ignore movimentos de mouse e cliques quando o `m_bOver != 1`. Sendo assim, tanto o código original, quanto o código descompilado quanto este fix parecem incorretos.

Minha sugestão seria:
1.  Fazer a mudança para ignorar cliques quando `m_bOver != 1`, bem parecido com o que você fez. Eu digo mudar para != 1 porque é o que normalmente o código faz.
2. Juntar o if do dwFlags != 513 com o if logo acima. sendo assim, sempre que for um clique, só aceitamos quando for no checkbox.

Sem o número dois, faça o seguinte teste. Insira o CheckBox numa janela qualquer e deixe aberto. Tente clicar em qualquer outro botão da janela, não vai funcionar o clique. Fechando a janela que possui o CheckBox, vai funcionar. O código final seria:
```c++
	if (dwFlags != 513 || m_bOver != 1)
		return SPanel::OnMouseEvent(dwFlags, wParam, nX, nY);
```

kevinkouketsu replied:
Sobre o clique que é muito rápido, infelizmente a responsabilidade do tempo de clique é do controle em si. Como você pode ver no SButton, por exemplo, no OnMouseEvernt, ele mesmo tem a responsabilidade de checar o tempo para não clicar muito rapidamente. Você pode ver isto aqui:
```c++
			unsigned int dwServerTime = g_pTimerManager->GetServerTime();
			if (dwServerTime > m_dwOldTime + 500)
			{
```

O ideal seria fazer o mesmo no SCheckBox.

seitbnao commented:
Oi, vou realizar as alterações, sim havia notado a questão do tempo no SPanel, segue um gif testando o código

![ezgif-4-082d124846](https://user-images.githubusercontent.com/11374956/182725800-0bb5fde3-08b6-4353-81fe-f31fcbc3a111.gif)
.

seitbnao commented:
> Gostaria de entender um pouco melhor. Como posso ver o problema acontecendo? O que devo notar?
Um único click, tende a clicar várias vezes, e em alguns momentos apenas passando o mouse encima o check, acaba trocando a textura entre on/off.

kevinkouketsu commented:
> > Gostaria de entender um pouco melhor. Como posso ver o problema acontecendo? O que devo notar?
> > Um único click, tende a clicar várias vezes, e em alguns momentos apenas passando o mouse encima o check, acaba trocando a textura entre on/off.

Eu consegui ver aqui. Acabei não comentando que havia checado, perdão!
De qualquer forma, acredito que com minhas sugestões o problema será solucionado como um todo. Se tiveres alguma coisa a acrescentar, let me know

seitbnao committed:
fix: checkbox, requested changes

seitbnao committed:
Revert "fix: checkbox, requested changes"

This reverts commit 40a89e7db958846c1690da6d6ca23b1347153426.

seitbnao committed:
Revert "Revert "fix: checkbox, requested changes""

This reverts commit 0825e9ff67f0f67d4f7f6036a57aa1e0bdb9dfd3.

seitbnao committed:
fix: checkbox, requested changes

Analisando melhor, cheguei a conclusão de que a SCheckBox herda a classe SButton, com estas alterações, a funcionalidade fica de forma correta.

kevinkouketsu commented:
Eu vou fazer alguns testes para ter certeza das implicações da alteração da herança antes de merjar esse pull request.

seitbnao commented:
> Eu vou fazer alguns testes para ter certeza das implicações da alteração da herança antes de merjar esse pull request.

Tudo bem hehe, aguardo ressalvas.
